Users can view issues not featured in the main sidebar

beausmith opened the pull request:
@oderby - Please review this PR. The only thing I didn't get to is renaming various "issues" files to "activeIssues" files. 

**Note** - This issue requires new endpoint that @oderby working on.

nickoneill commented:
I checked this with @oderby's backend changes, it's an excellent start. I can get the issue details to show up when directly linking but the list of issues both on the main / more issues page don't show anything. It looks like the issues are loading correctly (during the merge phase there are items in active and inactive lists) but not displayed? No console errors.

Some other things I noticed:
* location spins forever on the all issues page
* it looks like you added issue caching in here too but I didn't see a spot where it's invalidated if the location changes (i.e. new contacts, which are attached directly to issues as of now)

oderby commented on static/js/main.js:
```diff
@@ -105,14 +110,30 @@ app.model({
+    mergeIssues: (state, data) => {
+      issues = state.activeIssues.concat(state.inactiveIssues)
+      contactIndices = state.contactIndices
+      issues.map(issue => {
```
You should use `forEach` instead of `map` here, since this is purely side-effects.

beausmith replied:
Cool.

oderby commented on static/js/main.js:
```diff
@@ -105,14 +110,30 @@ app.model({
+      issues = state.activeIssues.concat(state.inactiveIssues)
+      contactIndices = state.contactIndices
+      issues.map(issue => {
+        contactIndices[issue.id] = contactIndices[issue.id] || 0
```
Why the change to keep old contactIndices? If I understand it, the previous code would always reset the contact indices. Now we won't. This change means that if someone changes their location after making it part way through an issue, they won't see the new contacts. There's probably some other complexities here, but unless there's a compelling reason to, I don't think we should change this behavior in this PR.

beausmith replied:
Reverted. I thought this was for another purpose, now I see what it's really for. Reverting.

nickoneill commented:
🏓

beausmith commented:
Did a quick look into the issues @nickoneill reported, was able to repro, but wasn't able to fix. Will look into it again tonight/tomorrow. If someone looks into it sooner, it's all good.

nickoneill commented:
Fixed this issue on the saucelabs side, next time it runs should be OK.

oderby commented on static/js/main.js:
```diff
@@ -188,23 +209,48 @@ app.model({
-        address += state.geolocation
+    fetchActiveIssues: (state, data, send, done) => {
+      if (((Date.now() - Number(state.lastActiveIssueFetchDate)) / 1000 / 60) < issueCacheMinutes && !!state.activeIssues.length) {
+        return
```
So looking at this with fresh eyes, what benefits do we get from manually caching issues on the client? With #213, the browser should cache the request for us automatically, and the merge logic is trivial and should run more than fast enough. Until that lands, it'd mean we'll fetch inactive issues the first time they load an inactive issue and each time thereafter they load the `/issues` page. I wonder if a slow-loading `/issues/` page is worth the reduction in code complexity? 

We could even go further and only fetch (in)active issues once, (by checking here and in `fetchInactiveIssues` if any (in)active issues are present in `state.issues`). This would mean they wouldn't see newly-added or changed (active<->inactive) issues without refreshing the page, but that seems like a really low-frequency event.

beausmith replied:
@oderby - I feel ya. 

Because #213 isn't done… should we make a note there to remove this manual caching? …or should I remove it now in anticipation of that task being completed?

I had thought about using the existence of (in)active issues as the cache control, but then thought that this solution would lead to less end-user confusion. 

Ok, I have removed the caching in a separate commit (https://github.com/5calls/5calls/pull/220/commits/ce7271d9b12108aac5e70ddfa27273b96986a574) in the case that someone else wants it back. Either way, I think this PR should be ready to roll.

beausmith commented:
Ok! @nickoneill I have fixed the issue you reported regarding loading the "more issues" page manually. I've also updated all tests to pass. Yay tests!

oderby commented on static/js/components/call.js:
```diff
@@ -9,9 +9,11 @@ const promote = require('./promote.js');
   if (issue == null) {
-    return html`<section class="call" onload=${(e) => send('oldcall')}>
+    return html`<section class="call" onload=${(e) => {
+      send('fetchInactiveIssues')
```
This feels a little hacky, but it's not incorrect and, for most people, will only result in a quick flash of "bad issue" text before we load the inactive issues list and display the issue content. If we start seeing people on slow connections or this becomes a problem, we can add an intermediate text saying "loading your issue, hold tight" or such.

nickoneill replied:
Same thing occurs today when linking to an issue, so I don't think this is a huge deal.

oderby replied:
:+1:

oderby commented on static/js/components/issues.js:
```diff
@@ -12,10 +13,15 @@ module.exports = (state, prev, send) => {
   }
 
+  function scrollToTop () {
+    scrollIntoView(document.querySelector('#content'))
```
Will this need to change with #226?

bernardwang replied:
It shouldn't! The function scrollIntoView has been redefined, but it is used exactly the same.

oderby commented on static/js/main.js:
```diff
@@ -321,7 +354,7 @@ app.model({
         }
```
Line 346 above should probably just look at `state.activeIssues.length`, right? I don't think it matters in practice, but there's technically a race between that check and the `fetchInactiveIssues` action and ensuing `mergeIssues` actions completing.

nickoneill replied:
I'm looking at line numbers and I'm not sure I understand where you're referencing here.

oderby replied:
yeah, it sucks that github doesn't let you comment on lines not within the original diff context... I'm referring to:
```javascript
       if (state.issues.length == 0 || state.geolocation == '') {
```
and suggesting it should be changed in this PR to read
```javascript
       if (state.activeIssues.length == 0 || state.geolocation == '') {
```

nickoneill commented:
Looks great to me! I'd like to make sure @oderby is good on that last comment he had but otherwise ready to go out.

Seeing the list was making Rebecca think about some more organization on that page, particularly if it starts to have too many inactive issues. Maybe some categories in Airtable are in order (in the future).

nickoneill commented:
Thank you @beausmith and @oderby!
